lib/scholar/cluster/hierarchical.ex

```diff
@@ -0,0 +1,476 @@
+      while {{clades, diss, sizes}, {count = 0, pointers, pairwise}}, count < n - 1 do
+        # Indexes of who I am nearest to
+        nearest = Nx.argmin(pairwise, axis: 1)
+
+        # Take who I am nearest to is nearest to
+        nearest_of_nearest = Nx.take(nearest, nearest)
+
+        # If the entry is pointing back at me, then we are a clade
+        clades_selector = Nx.equal(nearest_of_nearest, Nx.iota({n}))
+
+        # Now let's get the links that form clades.
+        # They are bidirectional but let's keep only one side.
+        links = Nx.select(clades_selector and nearest > nearest_of_nearest, nearest, n)
+
+        {clades, count, pointers, pairwise, diss, sizes} =
+          merge_clades(clades, count, pointers, pairwise, diss, sizes, links, n, update_fun)
+
+        {{clades, diss, sizes}, {count, pointers, pairwise}}
+      end
+
+    sizes = sizes[n..(2 * n - 2)]
+    perm = Nx.argsort(diss)
+    {clades[perm], diss[perm], sizes[perm]}
+  end
+
+  defnp merge_clades(clades, count, pointers, pairwise, diss, sizes, links, n, update_fun) do
+    {{clades, count, pointers, pairwise, diss, sizes}, _} =
+      while {{clades, count, pointers, pairwise, diss, sizes}, links},
+            i <- 0..(Nx.size(links) - 1) do
+        # i < j because of how links is formed.
+        # i will become the new clade index and we "infinity-out" j.
+        j = links[i]
+
+        if j == n do
+          {{clades, count, pointers, pairwise, diss, sizes}, links}
+        else
+          # Clades a and b (i and j of pairwise) are being merged into c.
+          a = find_clade(pointers, i)
+          b = find_clade(pointers, j)
+          c = count + n
+
+          # Update clades
+          new_clade = Nx.stack([a, b]) |> Nx.sort() |> Nx.new_axis(0)
+          clades = Nx.put_slice(clades, [count, 0], new_clade)
+
+          # Update sizes
+          sc = sizes[i] + sizes[j]
+          sizes = Nx.indexed_put(sizes, Nx.stack([i, c]) |> Nx.new_axis(-1), Nx.stack([sc, sc]))
+
+          # Update dissimilarities
+          diss = Nx.indexed_put(diss, Nx.stack([count]), pairwise[i][j])
+
+          # Update pairwise
+          {pairwise, _} =
+            while {pairwise, {x = i, y = j, sa = sizes[i], sb = sizes[j], sc = sc}},
+                  z <- 0..(n - 1) do
+              if z == x or z == y or Nx.is_infinity(pairwise[[0, z]]) do
+                {pairwise, {x, y, sa, sb, sc}}
+              else
+                dac = pairwise[[x, z]]
+                dbc = pairwise[[y, z]]
+                dab = pairwise[[x, y]]
+                update = update_fun.(dac, dbc, dab, sa, sb, sc)
+
+                # TODO: do this in a single call?
+                pairwise =
+                  pairwise
+                  |> Nx.indexed_put(Nx.stack([x, z]), update)
+                  |> Nx.indexed_put(Nx.stack([z, x]), update)
+                  |> Nx.indexed_put(Nx.stack([y, z]), update)
+                  |> Nx.indexed_put(Nx.stack([z, y]), update)
```

Comment:
Yeah, I think you can do a Nx.concatenate of the stacks and of the updates.

Reply:
Thanks! I _think_ I got it now: [35d0ebe](https://github.com/elixir-nx/scholar/pull/187/commits/35d0ebe8df7b9fbf77be2bb7f8b304db5cae4522). (But IDK if it's actually more performant...)

Reply:
I am not sure if it is more performant either, so go with whatever reads better (I prefer the version with several indexed_put, there are fewer operations, your call).